packages/did-resolver/src/base-resolver.ts

```diff
@@ -0,0 +1,92 @@
+import * as crypto from '@atproto/crypto'
+import { check } from '@atproto/common-web'
+import { AtprotoData, DidDocument, didDocument } from './types'
+import * as atprotoData from './atproto-data'
+import { DidNotFoundError, PoorlyFormattedDidDocumentError } from './errors'
+import { DidCache } from './did-cache'
+
+export abstract class BaseResolver {
+  constructor(public cache?: DidCache) {}
+
+  abstract resolveDidNoCheck(did: string): Promise<unknown | null>
+
+  validateDidDoc(did: string, val: unknown): DidDocument {
+    if (!check.is(val, didDocument)) {
+      throw new PoorlyFormattedDidDocumentError(did, val)
+    }
+    if (val.id !== did) {
+      throw new PoorlyFormattedDidDocumentError(did, val)
+    }
+    return val
+  }
+
+  async resolveDidNoCache(did: string): Promise<DidDocument | null> {
+    const got = await this.resolveDidNoCheck(did)
+    if (got === null) return null
+    return this.validateDidDoc(did, got)
+  }
+
+  async refreshCache(did: string): Promise<void> {
+    await this.cache?.refreshCache(did, () => this.resolveDidNoCache(did))
+  }
+
+  async resolveDid(
+    did: string,
+    forceRefresh = false,
+  ): Promise<DidDocument | null> {
+    if (this.cache && !forceRefresh) {
+      const fromCache = await this.cache.checkCache(did)
+      if (fromCache?.expired) {
+        this.refreshCache(did)
+      }
+      if (fromCache) {
+        return fromCache.doc
+      }
+    }
+
+    const got = await this.resolveDidNoCache(did)
+    if (got === null) return null
+    await this.cache?.cacheDid(did, got)
+    return got
```

Comment:
I could imagine a world in which a missing doc is cacheable.  I may need to think it through a bit more though, hmm 🤔

Reply:
I could see that... & not opposed to adding that in if we find it'd be helpful. My thought is that if you cached the missing doc & are asking to resolve the did, basically everytime you'll want to bypass the cache (cause you probably wouldn't be resolving if you thought it did not exist)